File: src/main/java/com/vowser/backend/api/controller/SpeechController.java
package com.vowser.backend.api.controller;

import com.vowser.backend.api.doc.SpeechApiDocument;
import com.vowser.backend.api.dto.speech.SpeechResponse;
import com.vowser.backend.api.dto.speech.SpeechTranscribeRequest;
import com.vowser.backend.application.service.speech.SpeechProcessingService;
import com.vowser.backend.common.constants.ApiConstants;
import com.vowser.backend.infrastructure.security.CustomUserDetails;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

@Slf4j
@Tag(name = "Speech Processing", description = "음성 인식 및 처리 API")
@RestController
@RequestMapping(ApiConstants.API_PATH_SPEECH)
@RequiredArgsConstructor
public class SpeechController {

    private final SpeechProcessingService speechProcessingService;

    @SpeechApiDocument.TranscribeAndExecute
    @PostMapping(value = "/transcribe", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    public ResponseEntity<SpeechResponse> transcribeAndExecute(
            @Valid @ModelAttribute SpeechTranscribeRequest request,
            @AuthenticationPrincipal CustomUserDetails userDetails) {

        // userDetails는 JWT 토큰이 유효하면 주입되고, 없으면 null이 됨
        return speechProcessingService.processVoiceCommand(request, userDetails);
    }

    @SpeechApiDocument.McpStatus
    @GetMapping("/mcp-status")
    public ResponseEntity<Object> getMcpConnectionStatus() {
        return speechProcessingService.getMcpConnectionStatus();
    }
}

File: src/main/kotlin/com/vowser/backend/api/controller/SpeechController.kt
package com.vowser.backend.api.controller

import com.vowser.backend.api.dto.SpeechResponse
import com.vowser.backend.application.service.SpeechService
import com.vowser.backend.infrastructure.mcp.McpWebSocketClient
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile

@RestController
@RequestMapping("/api/v1/speech")
class SpeechController(
    private val speechService: SpeechService,
    private val mcpWebSocketClient: McpWebSocketClient
) {
    private val logger = LoggerFactory.getLogger(javaClass)

    @PostMapping("/transcribe")
    fun transcribeAndExecute(
        @RequestParam("audio") audioFile: MultipartFile,
        @RequestParam("sessionId") sessionId: String
    ): ResponseEntity<SpeechResponse> {
        return try {
            val transcript = speechService.transcribe(audioFile)
            logger.info("음성 인식 완료: sessionId=[${sessionId}], transcript=[${transcript}]")

            if (mcpWebSocketClient.isConnected()) {
                mcpWebSocketClient.sendVoiceCommand(transcript, sessionId)
                logger.info("MCP 서버로 음성 명령 전송 완료")
            } else {
                logger.error("MCP 서버에 연결되지 않음")
                return ResponseEntity.internalServerError().body(
                    SpeechResponse(success = false, transcript = transcript, message = "MCP 서버에 연결되지 않음")
                )
            }

            ResponseEntity.ok(SpeechResponse(success = true, transcript = transcript))
        } catch (e: Exception) {
            logger.error("음성 처리 중 오류 발생", e)
            ResponseEntity.internalServerError().body(SpeechResponse(success = false, transcript = null, message = e.message))
        }
    }

}